src/Shared/BuildTransferredException.cs

```diff
@@ -0,0 +1,85 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.IO;
+
+namespace Microsoft.Build.BackEnd
+{
+    internal sealed class BuildTransferredException : Exception
+    {
+        private readonly string? _typeName;
+
+        public BuildTransferredException(
+            string? message,
+            Exception? inner,
+            string? typeName,
+            string deserializedStackTrace)
+            : base(message, inner)
+        {
+            _typeName = typeName;
+            StackTrace = deserializedStackTrace;
+        }
+
+        public override string? StackTrace { get; }
+
+        public override string ToString() => $"{_typeName ?? "Unknown"}->{base.ToString()}";
+
+        internal static Exception ReadExceptionFromTranslator(ITranslator translator)
+        {
+            BinaryReader reader = translator.Reader;
+            Exception? innerException = null;
+            if (reader.ReadBoolean())
+            {
+                innerException = ReadExceptionFromTranslator(translator);
+            }
+
+            string? message = ReadOptionalString(reader);
+            string? typeName = ReadOptionalString(reader);
+            string deserializedStackTrace = reader.ReadString();
+            BuildTransferredException exception = new(message, innerException, typeName, deserializedStackTrace)
+            {
+                Source = ReadOptionalString(reader),
+                HelpLink = ReadOptionalString(reader),
+                // HResult = reader.ReadInt32(),
+            };
+
+            return exception;
+        }
+
+        internal static void WriteExceptionToTranslator(ITranslator translator, Exception exception)
+        {
+            BinaryWriter writer = translator.Writer;
+            writer.Write(exception.InnerException != null);
+            if (exception.InnerException != null)
+            {
+                WriteExceptionToTranslator(translator, exception.InnerException);
+            }
+            WriteOptionalString(writer, exception.Message);
+            WriteOptionalString(writer, exception.GetType().FullName);
+            writer.Write(exception.StackTrace ?? string.Empty);
+            WriteOptionalString(writer, exception.Source);
+            WriteOptionalString(writer, exception.HelpLink);
+            // HResult is completely protected up till net4.5
+            // writer.Write(System.Runtime.InteropServices.Marshal.GetHRForException(exception));
+        }
+
+        private static string? ReadOptionalString(BinaryReader reader)
```

Comment:
This method already exists, albeit not here...think maybe you could unify the implementations?

Reply:
tl;dr;: true, but out of scope.

The `ReadOptionalString`, `WriteOptionalString` helpers are part of `Microsoft.Build.Framework assembly`, which makes them unavailable for `MSBuildTaskHost` assembly.
I'd need to extract them to a shared file that is build imported by each project separately - this makes developer life a misery and I do not want to spread this technique any further :-) However I do not have capacity to tackle this more systematically as part of this PR. So I decide to just copy those few lines of code

Reply:
I'll actually try to use technique suggested by @rainersigwald below (common utils in FW, linked into TaskHost) - hold on :-)